### DISCUSSION TITLE
Tips for preserving the data type of ref data after saving model (pickled data)

### DISCUSSION THREAD

**Discussion by happyych** [fumitoh/modelx#107] [Q&A]
Hi, I am really excited to test various modelx features nowadays.

I want to ask some tips about preserving the data type of ref data ( which I get via  new_pandas  function) after saving the model.

Below is an example which I faced some trouble.

1) I make ref variable in my model by using the code below

```python
df = pd.read_csv('test.csv', dtype = {'PROD_CODE' : 'string', 'PLAN_CODE' : 'string'}, index_col=['PROD_CODE', 'PLAN_CODE'])
P.new_pandas("test","test.csv",data=df,file_type ="csv",sheet = "sheet1")
```  
**sample data**

```
PROD_CODE PLAN_CODE RATE
"1111","0055",0.04
"1112","0001",0.03
```

2) I can get exact same data when I make the ref varable "test"

3) However, after I saved the model, it seems the data type I assigned to each column ( STRING to "PROD_CODE" and "PLAN_CODE" disappear)

--> When I reload the model, the data changed as below

**after reloading the model**

```
PROD_CODE PLAN_CODE RATE
1111,55,0.04
1112,1,0.03
```

If there is any way to preserve the data type of each column I first assigned, pleas let me know.

Thank you for reading my question!

**Comment by fumitoh** [comment]
It appears there's a bug in modelx where it fails to correctly restore the data types (dtypes) of string columns that only contain numeric values. I'll create an issue to address this.

In the meantime, there's a workaround you can use. Instead of directly storing your file within the model, you can place it externally and read it into the model. With the release of modelx 0.25.0, it's now possible to determine the path of the model where it was loaded from or saved to.

For instance, let’s say you have a data file named `test.csv` saved at `C:\Users\xxx\yyy\test.csv`. Here's how you can read it in your model:

```python
import modelx as mx
import pandas as pd

m = mx.new_model()

@mx.defcells
def df():
    file_path = _model.path.parent / 'test.csv'
    return pd.read_csv(file_path, dtype={'PROD_CODE': 'string', 'PLAN_CODE': 'string'}, index_col=['PROD_CODE', 'PLAN_CODE'])

m.write(r"C:\Users\xxx\yyy\TempModel")  # Sets Model.path to C:\Users\xxx\yyy\TempModel

print(df())  # Outputs the DataFrame with the correct dtypes
```

In this code, `_model.path.parent` retrieves the parent directory of the model's path (in this case, `C:\Users\xxx\yyy`) as a `pathlib.Path` object. For more details, refer to the [modelx path documentation](https://docs.modelx.io/en/latest/reference/generated/modelx.core.model.Model.path.html#modelx.core.model.Model.path).

**Reply by happyych** [reply]
Thank you for the kind answer!
I will try your suggestion right away.
Thanks!
